src/compiler/checker.ts

```diff
@@ -11069,66 +11127,145 @@ namespace ts {
+        /**
+         * Get attributes symbol of the given Jsx opening-like element. The result is from resolving "attributes" property of the opening-like element.
+         * @param openingLikeElement a Jsx opening-like element
+         * @return a symbol table resulted from resolving "attributes" property or undefined if any of the attribute resolved to any or there is no attributes.
+         */
+        function getJsxAttributesSymbolArrayFromAttributesProperty(openingLikeElement: JsxOpeningLikeElement): Symbol[] | undefined {
+            const attributes = openingLikeElement.attributes;
+            let attributesTable = createMap<Symbol>();
+            let spread: Type = emptyObjectType
+            let attributesArray: Symbol[] = [];
+            for (const attributeDecl of attributes.properties) {
+                const member = attributeDecl.symbol;
+                if (isJsxAttribute(attributeDecl)) {
+                    let exprType: Type;
+                    if (attributeDecl.initializer) {
+                        exprType = checkExpression(attributeDecl.initializer);
                     }
                     else {
-                        // If there's no corresponding property with this name, error
-                        if (!correspondingPropType) {
-                            error(node.name, Diagnostics.Property_0_does_not_exist_on_type_1, node.name.text, typeToString(elementAttributesType));
-                            return unknownType;
-                        }
+                        // <Elem attr /> is sugar for <Elem attr={true} />
+                        exprType = trueType;
+                    }
+
+                    const attributeSymbol = <TransientSymbol>createSymbol(SymbolFlags.Property | SymbolFlags.Transient | member.flags, member.name);
+                    attributeSymbol.declarations = member.declarations;
+                    attributeSymbol.parent = member.parent;
+                    if (member.valueDeclaration) {
+                        attributeSymbol.valueDeclaration = member.valueDeclaration;
                     }
+                    attributeSymbol.type = exprType;
+                    attributeSymbol.target = member;
+                    attributesTable[attributeSymbol.name] = attributeSymbol;
+                    attributesArray.push(attributeSymbol);
+                }
+                else {
+                    Debug.assert(attributeDecl.kind === SyntaxKind.JsxSpreadAttribute);
+                    if (attributesArray.length > 0) {
+                        spread = getSpreadType(spread, createJsxAttributesType(attributes.symbol, attributesTable), attributes.symbol);
+                        attributesArray = [];
+                        attributesTable = createMap<Symbol>();
+                    }
+                    const exprType = checkExpression(attributeDecl.expression);
+                    const widenExprType = getWidenedType(exprType);

```

Comment:
After reviewing this, I agree. I shouldn't widen here since if it is "undefined" then we shouldn't allow it to be spread.